File: src/lib/autosleep.ts
import net from "net";
import dgram from "dgram";
import { listServers, getServer } from "./servers";
import { getContainerInfo, startContainer, stopContainer } from "./docker";
import { queryServer } from "./a2s";

const IDLE_TIMEOUT_MS = 5 * 60 * 1000; // 5 minutes
const POLL_INTERVAL_MS = 15 * 1000; // check every 15 seconds

interface SleepState {
  lastSeenPlayers: number;
  emptyTimestamp: number | null; // when the server was first seen empty
  sleeping: boolean;
  wakeListener: { tcp: net.Server; udp: dgram.Socket } | null;
  waking: boolean; // currently booting up
}

const state = new Map<string, SleepState>();
let pollTimer: ReturnType<typeof setInterval> | null = null;

function getState(serverId: string): SleepState {
  if (!state.has(serverId)) {
    state.set(serverId, {
      lastSeenPlayers: -1,
      emptyTimestamp: null,
      sleeping: false,
      wakeListener: null,
      waking: false,
    });
  }
  return state.get(serverId)!;
}

export function isServerSleeping(serverId: string): boolean {
  return getState(serverId).sleeping;
}

export function isServerWaking(serverId: string): boolean {
  return getState(serverId).waking;
}

async function pollServers() {
  const servers = listServers();

  for (const server of servers) {
    if (!server.container_id) continue;

    const s = getState(server.id);

    // Skip if currently waking up
    if (s.waking) continue;

    // If sleeping, the wake listener handles it
    if (s.sleeping) continue;

    // Check if container is running
    const info = await getContainerInfo(server.container_id);
    if (!info || info.state !== "running") continue;

    // Query player count via RCON
    const result = await queryServer("127.0.0.1", server.port);
    const players = result?.players ?? -1;

    if (players === -1) {
      // RCON failed, server might still be booting — skip
      continue;
    }

    if (players > 0) {
      // Players online — reset idle timer
      s.lastSeenPlayers = players;
      s.emptyTimestamp = null;
    } else {
      // Empty server
      if (s.emptyTimestamp === null) {
        s.emptyTimestamp = Date.now();
        console.log(`[autosleep] ${server.name} (port ${server.port}): empty, starting idle timer`);
      } else if (Date.now() - s.emptyTimestamp >= IDLE_TIMEOUT_MS) {
        // Idle timeout reached — put server to sleep
        console.log(`[autosleep] ${server.name} (port ${server.port}): idle for 5 min, sleeping`);
        await sleepServer(server.id, server.port, server.container_id);
      }
    }
  }
}

async function sleepServer(serverId: string, port: number, containerId: string) {
  const s = getState(serverId);

  try {
    // Stop the container (don't remove — just stop)
    await stopContainer(containerId);
  } catch (err) {
    console.error(`[autosleep] Failed to stop container:`, err);
    return;
  }

  s.sleeping = true;
  s.emptyTimestamp = null;

  // Start wake listeners on the server's port
  startWakeListener(serverId, port, containerId);
}

function startWakeListener(serverId: string, port: number, containerId: string) {
  const s = getState(serverId);

  // TCP listener
  const tcp = net.createServer((socket) => {
    console.log(`[autosleep] TCP connection on port ${port} — waking server ${serverId}`);
    socket.destroy();
    wakeServer(serverId, port, containerId);
  });

  tcp.on("error", (err: NodeJS.ErrnoException) => {
    if (err.code === "EADDRINUSE") {
      console.log(`[autosleep] Port ${port} TCP still in use, retrying in 5s`);
      setTimeout(() => {
        tcp.close();
        startWakeListener(serverId, port, containerId);
      }, 5000);
    }
  });

  // UDP listener
  const udp = dgram.createSocket("udp4");

  udp.on("message", () => {
    console.log(`[autosleep] UDP packet on port ${port} — waking server ${serverId}`);
    wakeServer(serverId, port, containerId);
  });

  udp.on("error", (err: NodeJS.ErrnoException) => {
    if (err.code === "EADDRINUSE") {
      console.log(`[autosleep] Port ${port} UDP still in use, retrying in 5s`);
      setTimeout(() => {
        udp.close();
        startWakeListener(serverId, port, containerId);
      }, 5000);
    }
  });

  // Wait a moment for Docker to release the port, then bind
  setTimeout(() => {
    try {
      tcp.listen(port, "0.0.0.0", () => {
        console.log(`[autosleep] Wake listener TCP on port ${port}`);
      });
      udp.bind(port, "0.0.0.0", () => {
        console.log(`[autosleep] Wake listener UDP on port ${port}`);
      });
      s.wakeListener = { tcp, udp };
    } catch (err) {
      console.error(`[autosleep] Failed to start wake listener:`, err);
    }
  }, 3000);
}

async function wakeServer(serverId: string, port: number, containerId: string) {
  const s = getState(serverId);

  // Prevent multiple simultaneous wakes
  if (s.waking) return;
  s.waking = true;

  // Close wake listeners first to free the port
  if (s.wakeListener) {
    s.wakeListener.tcp.close();
    s.wakeListener.udp.close();
    s.wakeListener = null;
  }

  console.log(`[autosleep] Waking server ${serverId} on port ${port}`);

  // Small delay to ensure port is released
  await new Promise((r) => setTimeout(r, 1000));

  try {
    await startContainer(containerId);
    console.log(`[autosleep] Server ${serverId} started`);
  } catch (err) {
    console.error(`[autosleep] Failed to start container:`, err);
  }

  s.sleeping = false;
  s.waking = false;
  s.emptyTimestamp = null;
  s.lastSeenPlayers = -1;
}

// Manually wake a sleeping server (from the dashboard)
export async function manualWake(serverId: string) {
  const server = getServer(serverId);
  if (!server?.container_id) return;

  const s = getState(serverId);
  if (!s.sleeping) return;

  await wakeServer(serverId, server.port, server.container_id);
}

// Manually reset sleep state (e.g. when user manually starts/stops)
export function resetSleepState(serverId: string) {
  const s = getState(serverId);
  if (s.wakeListener) {
    s.wakeListener.tcp.close();
    s.wakeListener.udp.close();
    s.wakeListener = null;
  }
  s.sleeping = false;
  s.waking = false;
  s.emptyTimestamp = null;
  s.lastSeenPlayers = -1;
}

export function startAutoSleep() {
  if (pollTimer) return;
  console.log("[autosleep] Started — polling every 15s, idle timeout 5 min");
  pollTimer = setInterval(pollServers, POLL_INTERVAL_MS);
}

export function stopAutoSleep() {
  if (pollTimer) {
    clearInterval(pollTimer);
    pollTimer = null;
  }
  // Clean up all listeners
  for (const [, s] of state) {
    if (s.wakeListener) {
      s.wakeListener.tcp.close();
      s.wakeListener.udp.close();
    }
  }
  state.clear();
}

File: cli/internal/ddsm/autosleep.go
package ddsm

import (
	"context"
	"fmt"
	"net"
	"sync"
	"time"
)

type sleepState struct {
	emptyTimestamp *time.Time
	sleeping       bool
	waking         bool
	tcpListener    net.Listener
	udpConn        *net.UDPConn
	mu             sync.Mutex
}

var (
	sleepStates   = make(map[string]*sleepState)
	sleepStatesMu sync.Mutex
	sleepCancel   context.CancelFunc
	notifyFn      func(string)
	notifyMu      sync.RWMutex
)

// SetNotifier sets a function that receives autosleep log messages.
// When running inside the TUI, this routes messages through bubbletea
// instead of writing directly to stdout (which corrupts the display).
func SetNotifier(fn func(string)) {
	notifyMu.Lock()
	defer notifyMu.Unlock()
	notifyFn = fn
}

func notify(format string, args ...interface{}) {
	msg := fmt.Sprintf(format, args...)
	notifyMu.RLock()
	fn := notifyFn
	notifyMu.RUnlock()
	if fn != nil {
		fn(msg)
	}
}

func getSleepState(serverID string) *sleepState {
	sleepStatesMu.Lock()
	defer sleepStatesMu.Unlock()
	if s, ok := sleepStates[serverID]; ok {
		return s
	}
	s := &sleepState{}
	sleepStates[serverID] = s
	return s
}

func IsSleeping(serverID string) bool {
	s := getSleepState(serverID)
	s.mu.Lock()
	defer s.mu.Unlock()
	return s.sleeping
}

func IsWaking(serverID string) bool {
	s := getSleepState(serverID)
	s.mu.Lock()
	defer s.mu.Unlock()
	return s.waking
}

func StartAutoSleep() {
	if !Cfg.AutoSleep.Enabled {
		return
	}

	ctx, cancel := context.WithCancel(context.Background())
	sleepCancel = cancel

	interval := time.Duration(Cfg.AutoSleep.PollInterval) * time.Second
	idleTimeout := time.Duration(Cfg.AutoSleep.IdleTimeout) * time.Second

	notify("[autosleep] Started — polling every %ds, idle timeout %ds", Cfg.AutoSleep.PollInterval, Cfg.AutoSleep.IdleTimeout)

	go func() {
		ticker := time.NewTicker(interval)
		defer ticker.Stop()
		for {
			select {
			case <-ctx.Done():
				return
			case <-ticker.C:
				pollServers(idleTimeout)
			}
		}
	}()
}

func StopAutoSleep() {
	if sleepCancel != nil {
		sleepCancel()
	}
	sleepStatesMu.Lock()
	defer sleepStatesMu.Unlock()
	for _, s := range sleepStates {
		s.mu.Lock()
		closeListeners(s)
		s.mu.Unlock()
	}
	sleepStates = make(map[string]*sleepState)
}

func pollServers(idleTimeout time.Duration) {
	servers, err := ListServers()
	if err != nil {
		return
	}

	for _, server := range servers {
		if !server.ContainerID.Valid {
			continue
		}

		s := getSleepState(server.ID)
		s.mu.Lock()

		if s.waking || s.sleeping {
			s.mu.Unlock()
			continue
		}

		info, err := GetContainerInfo(server.ContainerID.String)
		if err != nil || info.State != "running" {
			s.mu.Unlock()
			continue
		}

		players, err := QueryServerPlayers(server.Port)
		if err != nil || players == nil {
			s.mu.Unlock()
			continue
		}

		if players.Players > 0 {
			s.emptyTimestamp = nil
		} else {
			if s.emptyTimestamp == nil {
				now := time.Now()
				s.emptyTimestamp = &now
				notify("[autosleep] %s (port %d): empty, starting idle timer", server.Name, server.Port)
			} else if time.Since(*s.emptyTimestamp) >= idleTimeout {
				notify("[autosleep] %s (port %d): idle for %ds, sleeping", server.Name, server.Port, int(idleTimeout.Seconds()))
				s.mu.Unlock()
				sleepServer(server.ID, server.Port, server.ContainerID.String)
				continue
			}
		}

		s.mu.Unlock()
	}
}

func sleepServer(serverID string, port int, containerID string) {
	if err := StopContainer(containerID); err != nil {
		notify("[autosleep] Failed to stop container: %v", err)
		return
	}

	s := getSleepState(serverID)
	s.mu.Lock()
	s.sleeping = true
	s.emptyTimestamp = nil
	s.mu.Unlock()

	time.Sleep(3 * time.Second)
	startWakeListener(serverID, port, containerID)
}

func startWakeListener(serverID string, port int, containerID string) {
	s := getSleepState(serverID)

	addr := fmt.Sprintf("0.0.0.0:%d", port)

	tcpListener, err := net.Listen("tcp", addr)
	if err != nil {
		notify("[autosleep] Port %d TCP in use, retrying in 5s", port)
		time.AfterFunc(5*time.Second, func() {
			startWakeListener(serverID, port, containerID)
		})
		return
	}

	udpAddr, _ := net.ResolveUDPAddr("udp", addr)
	udpConn, err := net.ListenUDP("udp", udpAddr)
	if err != nil {
		tcpListener.Close()
		notify("[autosleep] Port %d UDP in use, retrying in 5s", port)
		time.AfterFunc(5*time.Second, func() {
			startWakeListener(serverID, port, containerID)
		})
		return
	}

	s.mu.Lock()
	s.tcpListener = tcpListener
	s.udpConn = udpConn
	s.mu.Unlock()

	notify("[autosleep] Wake listener on port %d (TCP+UDP)", port)

	go func() {
		for {
			conn, err := tcpListener.Accept()
			if err != nil {
				return
			}
			conn.Close()
			notify("[autosleep] TCP connection on port %d — waking server", port)
			wakeServer(serverID, port, containerID)
			return
		}
	}()

	go func() {
		buf := make([]byte, 1)
		for {
			_, _, err := udpConn.ReadFromUDP(buf)
			if err != nil {
				return
			}
			notify("[autosleep] UDP packet on port %d — waking server", port)
			wakeServer(serverID, port, containerID)
			return
		}
	}()
}

func wakeServer(serverID string, port int, containerID string) {
	s := getSleepState(serverID)
	s.mu.Lock()
	if s.waking {
		s.mu.Unlock()
		return
	}
	s.waking = true
	closeListeners(s)
	s.mu.Unlock()

	notify("[autosleep] Waking server on port %d", port)

	time.Sleep(1 * time.Second)

	if UsesOverlay(serverID) {
		if err := MountOverlay(serverID); err != nil {
			notify("[autosleep] Failed to mount overlay: %v", err)
		}
	}

	if err := StartContainer(containerID); err != nil {
		notify("[autosleep] Failed to start container: %v", err)
	} else {
		notify("[autosleep] Server started on port %d", port)
	}

	s.mu.Lock()
	s.sleeping = false
	s.waking = false
	s.emptyTimestamp = nil
	s.mu.Unlock()
}

func ManualWake(serverID string) error {
	server, err := GetServer(serverID)
	if err != nil || server == nil {
		return fmt.Errorf("server not found: %s", serverID)
	}
	if !server.ContainerID.Valid {
		return fmt.Errorf("server has no container")
	}

	s := getSleepState(serverID)
	s.mu.Lock()
	if !s.sleeping {
		s.mu.Unlock()
		return fmt.Errorf("server is not sleeping")
	}
	s.mu.Unlock()

	wakeServer(serverID, server.Port, server.ContainerID.String)
	return nil
}

func ResetSleepState(serverID string) {
	s := getSleepState(serverID)
	s.mu.Lock()
	defer s.mu.Unlock()
	closeListeners(s)
	s.sleeping = false
	s.waking = false
	s.emptyTimestamp = nil
}

func closeListeners(s *sleepState) {
	if s.tcpListener != nil {
		s.tcpListener.Close()
		s.tcpListener = nil
	}
	if s.udpConn != nil {
		s.udpConn.Close()
		s.udpConn = nil
	}
}
